mypy/checkexpr.py

```diff
@@ -327,7 +327,20 @@ def infer_arg_types_in_context2(
         for i, actuals in enumerate(formal_to_actual):
             for ai in actuals:
                 if arg_kinds[ai] not in (nodes.ARG_STAR, nodes.ARG_STAR2):
-                    res[ai] = self.accept(args[ai], callee.arg_types[i])
+                    callee_arg_type = callee.arg_types[i]
+                    arg_type = self.accept(args[ai], callee_arg_type)
+                    if (isinstance(arg_type, CallableType) and
```

Comment:
I think you're talking about a case like
```
def f(xs: List[Callable[[T], S]], y: T) -> List[S]: pass
def id(x: U) -> U: pass

f([id], 1)
```
In principle we should be able to type check this as follows:

* The first pass of type checking finds that `T = int`.
* `f` provides the context `List[Callable[[int], S]]` to the expression `[id]`, which is itself treated as a call to a generic function "`[_]`" that produces a list.
* `[_]` provides the context `Callable[[int], S]` to the expression `id`. Now when `id` is type checked, the result is a generic callable type, so the code in this change unifies that type with the context, resulting in `Callable[[int], int]`.
* Now the first argument of `f` has type `List[Callable[[int], int]]` so we can infer that `S = int`.

In practice the third step doesn't work yet because we don't substitute variables in `infer_function_type_arguments_using_context` when the replacement would involve a type variable (here `S`).